src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -412,36 +417,64 @@ ShouldLazyDeparseQuery(Task *task)
 
 
 /*
- * SetTaskQuery attaches the query to the task so that it can be used during
- * execution. If local execution can possibly take place it sets task->queryForLocalExecution.
+ * SetTaskQueryIfShouldLazyDeparse attaches the query to the task so that it can be used during
+ * execution. If local execution can possibly take place it sets task->jobQueryReferenceForLazyDeparsing.
  * If not it deparses the query and sets queryStringLazy, to avoid blowing the
  * size of the task unnecesarily.
  */
 void
-SetTaskQuery(Task *task, Query *query)
+SetTaskQueryIfShouldLazyDeparse(Task *task, Query *query)
 {
 	if (ShouldLazyDeparseQuery(task))
 	{
-		task->queryForLocalExecution = query;
-		task->queryStringLazy = NULL;
+		InitializeTaskQueryIfNecessary(task);
+		task->taskQuery->queryType = TASK_QUERY_OBJECT;
+		task->taskQuery->data.jobQueryReferenceForLazyDeparsing = query;
 		return;
 	}
 
-	task->queryForLocalExecution = NULL;
-	task->queryStringLazy = DeparseTaskQuery(task, query);
+	SetTaskQueryString(task, DeparseTaskQuery(task, query));
 }
 
 
 /*
  * SetTaskQueryString attaches the query string to the task so that it can be
- * used during execution. It also unsets queryForLocalExecution to be sure
+ * used during execution. It also unsets jobQueryReferenceForLazyDeparsing to be sure
  * these are kept in sync.
  */
 void
 SetTaskQueryString(Task *task, char *queryString)
 {
-	task->queryForLocalExecution = NULL;
-	task->queryStringLazy = queryString;
+	InitializeTaskQueryIfNecessary(task);
+	task->taskQuery->queryType = TASK_QUERY_TEXT;
+	task->taskQuery->data.queryStringLazy = queryString;
+}
+
+
+/*
+ * InitializeTaskQueryIfNecessary initializes task query if it
+ * is not yet allocated.
+ */
+static void
+InitializeTaskQueryIfNecessary(Task *task)
+{
+	if (task->taskQuery == NULL)
+	{
+		task->taskQuery = CitusMakeNode(TaskQuery);
```

Comment:
Can we get into any situations where the Task sits in a long-lived memory context (e.g. prepared statement), but the task query does not?

Reply:
I guess here we should be using the task memory context, right? Because task query should live as long as the task.

So maybe:
```c
MemoryContext *previousContext = MemoryContextSwitchTo(GetMemoryChunkContext(
																   task));
task->taskQuery = CitusMakeNode(TaskQuery);
MemoryContextSwitchTo(previousContext);
```

Reply:
Could we just make taskQuery a struct rather than a pointer to a struct?

Reply:
What would be the advantage of that over using the same memory context as task?

Reply:
Simpler, no extra allocation, no chance of OOM.